src-java/ping-topology/ping-storm-topology/src/main/java/org/openkilda/wfm/topology/ping/bolt/PingProducer.java

```diff
@@ -66,20 +72,39 @@ private Ping buildPing(PingContext pingContext, FlowDirection direction) {
         Flow flow = pingContext.getFlow();
         FlowEndpoint ingress;
         FlowEndpoint egress;
+        int islPort;
         if (FlowDirection.FORWARD == direction) {
             ingress = new FlowSourceAdapter(flow).getEndpoint();
             egress = new FlowDestAdapter(flow).getEndpoint();
+            islPort = getIslPort(flow, flow.getForwardPath());
         } else if (FlowDirection.REVERSE == direction) {
             ingress = new FlowDestAdapter(flow).getEndpoint();
             egress = new FlowSourceAdapter(flow).getEndpoint();
+            islPort = getIslPort(flow, flow.getReversePath());
         } else {
             throw new IllegalArgumentException(String.format(
                     "Unexpected %s value: %s", FlowDirection.class.getCanonicalName(), direction));
         }
 
-        return new Ping(ingress.getOuterVlanId(), ingress.getInnerVlanId(),
-                        new NetworkEndpoint(ingress.getSwitchId(), ingress.getPortNumber()),
-                        new NetworkEndpoint(egress.getSwitchId(), egress.getPortNumber()));
+        return new Ping(new NetworkEndpoint(ingress.getSwitchId(), ingress.getPortNumber()),
+                        new NetworkEndpoint(egress.getSwitchId(), egress.getPortNumber()),
+                        pingContext.getTransitEncapsulation(), islPort);
+    }
+
+    private int getIslPort(Flow flow, FlowPath flowPath) {
+        List<PathSegment> segments = flowPath.getSegments();
+        if (segments.isEmpty()) {
+            throw new IllegalArgumentException(
+                    format("Path segments not provided, flow_id: %s", flow.getFlowId()));
+        }
+
+        PathSegment ingressSegment = segments.get(0);
+        if (!ingressSegment.getSrcSwitch().getSwitchId().equals(flowPath.getSrcSwitch().getSwitchId())) {
```

Comment:
Why this check is required?

Reply:
We verify this when creating requests for installing switch rules. Therefore, I think that there may be cases when this check is needed.